jaxopt/_src/levenberg_marquardt.py

```diff
@@ -0,0 +1,356 @@
+
+    tree_mul_term = tree_mul(self.xtol,
+                             tree_add(tree_l2_norm(params), self.xtol))
+    dparams_diff = tree_sub(tree_l2_norm(delta_params), tree_mul_term)
+    dparams_error = jax.lax.max(dparams_diff, 0.0)
+    error = jax.lax.min(state.error, dparams_error)
+
+    # Initializing the conditional input.
+    small_dparams_test_init_state = (params, error, state.grad_error,
+                                     state.damping_factor,
+                                     state.increase_factor, state.gradient,
+                                     state.residual)
+
+    # Checking if the dparams satisfy the "sufficiently small" criteria.
+    params, error, grad_error, damping_factor, increase_factor, gradient, residual = (
+        self.update_state_using_delta_params(loss_curr, delta_params,
+                                             dparams_diff,
+                                             small_dparams_test_init_state,
+                                             *args, **kwargs))
+
+    state = LevenbergMarquardtState(
+        iter_num=state.iter_num + 1,
+        damping_factor=damping_factor,
+        increase_factor=increase_factor,
+        error=error,
+        dparams_error=dparams_error,
+        grad_error=grad_error,
+        residual=residual,
+        loss=0.5 * jnp.dot(residual, residual),
+        delta=delta_params,
+        gradient=gradient)
+
+    return base.OptStep(params=params, state=state)
+
+  def __post_init__(self):
+    if self.has_aux:
+      self._fun = lambda *a, **kw: self.residual_fun(*a, **kw)[0]
+      self._fun_with_aux = self.fun
+    else:
+      self._fun = self.residual_fun
+      self._fun_with_aux = lambda *a, **kw: (self.residual_fun(*a, **kw), None)
+
+  def optimality_fun(self, params, *args, **kwargs):
+    """Optimality function mapping compatible with ``@custom_root``."""
+    residual = self._fun(params, *args, **kwargs)
+    return self._jt_op(params, residual, *args, **kwargs)
+
+  def _jt_op(self, params, residual, *args, **kwargs):
+    """Product of J^T and residual -- J: jacobian of fun at params."""
+    fun_with_args = lambda p: self._fun(p, *args, **kwargs)
+    _, vjpfun = jax.vjp(fun_with_args, params)
+    jt_op_val, = vjpfun(residual)
+    return jt_op_val
+
+  def _jtj_op(self, params, vec, *args, **kwargs):
+    """Product of J^T.J with vec using vjp & jvp, where J is jacobian of fun at params."""
+    fun_with_args = lambda p: self._fun(p, *args, **kwargs)
+    _, vjpfun = jax.vjp(fun_with_args, params)
+    _, jvp_val = jax.jvp(fun_with_args, (params,), (vec,))
+    jtj_op_val, = vjpfun(jvp_val)
+    return jtj_op_val
+
+  def _jtj_reg_op(self, params, lam, vec, *args, **kwargs):
+    """Product of (J^T.J + lam*I) and vec, where J is jacobian of fun at params."""
+    return self._jtj_op(params, vec, *args, **
+                        kwargs) + lam * jnp.dot(jnp.eye(len(params)), vec)
+
+  def _jtj_diag_op(self, params, *args, **kwargs):
```

Comment:
Unfortunately, I think this is just as expensive as computing the full Jacobian once for all and store it. I see that this information is used for updating the damping factor. Does it mean that the Levenberg-Marquardt algorithm is not "matvec" friendly? Maybe @shoyer has some thoughts on this.

Reply:
This is one way to initialize the damping factor, but it is pretty typical. Although, this is only once during the optimization.

Reply:
If it's done only once in the initialization, then that's fine! Sorry I thought it was done in `update`.

Reply:
no worries, yeah it is just for `init_state`.